### PR TITLE
#12840: Added unit tests for yii.validation.js

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes, it contains some bug fixes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #12840

Used @nkovacs draft as starting point(https://github.com/yiisoft/yii2/issues/11037#issuecomment-193356450). I also fixed some bugs in `yii.validation.js` and did some refactoring.

## Implemented tests

- yii.validation.js

## Running tests

- JS tests use jQuery and Punycode.js libraries from `vendor/bower` to match those that are used in assets and prevent installing and synchronizing duplicates via NPM. So make sure you have them installed. If no - run `composer install` to install them via Composer Asset Plugin.
- Install Node.js. I used the latest version available at the moment - `v7.1.0`. Note that used `jsdom` dependent library version requires `Node.js >= 4`.
- Run `npm install` to install dependencies.
- Execute `npm test` for running tests.

## Problems list

- [x] Sinon can not be installed via Composer Asset Plugin. If I add line `"npm-asset/sinon": "^1.17.6"` to `required-dev` section of `composer.json` and run `composer update` I get the following error:

```
[RuntimeException]
Failed to execute git clone --mirror 'git+ssh://git@github.com/sinonjs/sinon.git' '/home/vagrant/.composer/cache/vcs/git-ssh---git-github.com-sinonjs-sinon.git/'
Cloning into bare repository '/home/vagrant/.composer/cache/vcs/git-ssh---git-github.com-sinonjs-sinon.git'...
Permission denied (publickey).
fatal: Could not read from remote repository.
Please make sure you have the correct access rights and the repository exists.
```

Possibly related: https://github.com/fxpio/composer-asset-plugin/issues/187.

I tried `--prefer-dist` option, that didn't help. I also have generated SSH key and it's added it to `ssh-agent` (error is mentioned in [git-workflow](https://github.com/yiisoft/yii2/blob/master/docs/internals/git-workflow.md#1-fork-the-yii-repository-on-github-and-clone-your-fork-to-your-development-environment) document which is a part of internal docs).

This is not actual anymore because Node.js and NPM are used for testing.

- [x] Sometimes "image validator with FileReader API is not available" test takes a long time to load and gives warning with time written inside of a red label. This is because of swapping native `FileReader`:

```js
var initialFileReader = FileReader;

beforeEach(function () {
    FileReader = undefined;
});

afterEach(function () {
    FileReader = initialFileReader;
});
```

But:

- It happens pretty seldom and increases time by about half of second only.
- I can't see a better solution for testing this, so this is not a big problem.
- Similar technique is used for swapping native JS components in other places.

So this is not a big deal.

This is not actual anymore because Node.js and NPM are used for testing and for testing that case different approach is used.

- [x] Run tests in command line instead of a browser.

Node.js and NPM are used for testing.

- [x] Error while running `npm test`:

> Error: timeout of 15000ms exceeded. Ensure the done() callback is being called in this test.

Happens very seldom, this is because of jsdom and integrating jQuery. Default timeout is 2000ms. And this is always not enough, so I increased it to be 15000ms.

Solved with increasing timeout to 30s, see #13160.

- [x] In build detailed info the line after "with width more than minimum width" (validateImage method, with successfully read image) is not printed. It's OK locally though.

Can't see it in some builds, maybe it was Travis glitch related with formatting.

It's gone in the latest builds, for example in [this one](https://travis-ci.org/yiisoft/yii2/jobs/182190268).

## Integration with Travis CI

I found this [article](https://ma27.github.io/travis-ci/2016/05/06/travis-for-nodejs-and-php.html) and used it as starting point.

> TravisCI uses the software which is shipped by default on a `precise`: An ancient setup with NodeJS 0.10 and NPM 1.x. The current 3.x series of NPM is much more stable than the ancient NPM 1.x and fixed a lot of misbehavior which has been implemented in the older versions of NPM. Especially code which has been written for the browser will experience issues due to some implementation details of NPM releases prior to the 3.x series.

We can't use Node.js < 4 because of [jsdom](https://github.com/tmpvar/jsdom) dependency. I refused to use [3.x branch](https://github.com/tmpvar/jsdom/tree/3.x) of jsdom because it's considered legacy and no longer maintained.

I tried to change dist to `trusty` and install newer versions of Node.js and NPM. It worked, but another error appeared during creating test MySQL database:

```
- mysql -e 'CREATE DATABASE yiitest;';
```

Error's text:

> ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)

Then I found https://github.com/travis-ci/travis-ci/issues/6842 and fixed it with installation of MySQL 5.6 (default is MySQL 5.5 for trusty dist).

Hovewer new problems appeared.

- [x] On PHP 5.4:

> Warning: PHP Startup: Unable to load dynamic library
> '/home/travis/.phpenv/versions/5.4.45/lib/php/extensions/no-debug-zts-20100525/apc.so' -
> /home/travis/.phpenv/versions/5.4.45/lib/php/extensions/no-debug-zts-20100525/apc.so:
> cannot open shared object file: 
> No such file or directory in Unknown on line 0

Related: https://github.com/travis-ci/travis-ci/issues/6853

I'm trying to fix it by installing PECL package, see this [article](http://olvlvl.com/2016-09-travis-apc). Hovewer no results yet, because it tries to install it for PHP 7 version (that's weird because version in the build is 5.4).

Fixed.

- [x] On HHVM 2 tests related with PostgreSQL are now failing.

Specified 3.6 version for testing (the same as on Precise), created separate issues - #12981, #12982.

- [x] Get rid of oauth token, #12987.

### CODE REVIEW

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Fixed bug. 

In JavaScript array can not be determined just by using this kind of check: `value == []`, `[] == []` is always `false` for example. Since the code is jQuery dependent it's OK to use `$.isArray` (and by the way it's used above in the same file). Another advantage of using `$.isArray` over [Array.isArray()](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray) is that 1st version branch has fallback for older browsers, so in case user want that he need just override jquery version in `yii\web\JQueryAsset`.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: `is` conflicts with `min` and `max` and it does not make sense to use these options together. We either check for exact value or `min` or `max`. So it's better to skip `min` and `max` checks if `is` option is specified and check it first.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Inconsistency. There is no newline separator in similar code in file validator.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: It's now clear that this is an array.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Removed reduction. Moved `validateImage` to separate method because it's impossible to test this logic otherwise - local variables exist only in their scope and we can't redefine native event handlers. This for example:

```js
FileReader.prototype.onload = undefined;
```

gives:

```
TypeError: 'set onload' called on an object that does not implement interface FileReader.
```

I tried my best to think a workaround, but seems like moving it to separate method is the best solution to avoid ugly hacks.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Removed variable names reductions.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: We can use `this,result` instead in the scope of event handler.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: We expect client validators to behave the same way as according server validators. By default `not` is `false`, but if it was set to undefined, we get wrong result.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Extra newlines, variables can be declared together.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Code style. Missed semicolon.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: It's better to use separate named variables for local-part and domain because they are used more than once.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Code style. Can be refactored to one liner, does not exceed max string length.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Code style. Use the same way of declaring multiple variables at once.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Separate function is useless here because it's used only once and it's just one short line.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Better grouping of variables.

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Separate function is useless (see above).

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: What's the point of declaring `index` as `undefined` initially if then immediately assign some value to it?

File: framework/assets/yii.validation.js
Comment (arogachev) [inline framework/assets/yii.validation.js]: Good candidate for one liner.

File: framework/assets/yii.validation.js
Comment (dynasource) [inline framework/assets/yii.validation.js]: agreed

File: framework/assets/yii.validation.js
Comment (dynasource) [inline framework/assets/yii.validation.js]: agreed

File: framework/assets/yii.validation.js
Comment (dynasource) [inline framework/assets/yii.validation.js]: agreed

File: framework/assets/yii.validation.js
Comment (dynasource) [inline framework/assets/yii.validation.js]: agreed

**Comment by dynasource** [comment]
your PR is impressive. You've put a lot of effort in. Good job.

A critical remark though ;). Wasnt it possible to create the tests first and refactor yii.validation.js after? This would improve the trust on the refactorings. On the contrary, I can well imagine the requirement to modularize the JS first.

Regarding the other js file refactorings. Lets fix these 1 per PR (easier to review & revert if necessary)

**Review by dynasource — APPROVED** [review]
your PR is impressive. You've put a lot of effort in. Good job.

A critical remark though ;). Wasnt it possible to create the tests first and refactor yii.validation.js after? This would improve the trust on the refactorings. On the contrary, I can well imagine the requirement to modularize the JS first.

Regarding the other js file refactorings. Lets fix these 1 per PR (easier to review & revert if necessary)

**Comment by dynasource** [comment]
@arogachev,@cebe, it is already possible to execute the tests in Travis?

**Comment by arogachev** [comment]
@dynasource Thanks. Yes, refactoring can be done separately, but there are few bugs related to tests. Yes, I think covering one js file with tests per PR is better for review.

File: framework/assets/yii.validation.js
Comment (nkovacs) [inline framework/assets/yii.validation.js]: This is because in es3 boolean is a keyword, and could not be used without quotes.
It's ok in es5.

Comment (arogachev) [inline framework/assets/yii.validation.js]: @nkovacs Got it. Thanks for explanation. Should we leave it as is in this case to support ES3?

Comment (dynasource) [inline framework/assets/yii.validation.js]: @samdark?

Comment (nkovacs) [inline framework/assets/yii.validation.js]: Not IE9, IE8 and older.

Comment (arogachev) [inline framework/assets/yii.validation.js]: Got it. I'll return it as it was. I'll also add additional explaining comment about it in this case.

File: tests/js/assets/js/yii.validation.test.js
Comment (arogachev) [inline tests/js/assets/js/yii.validation.test.js]: I think my variation is more specific. With this message it's unclear what kind of result we actually expect from given data set.

Comment (nkovacs) [inline tests/js/assets/js/yii.validation.test.js]: Why is it unclear?

The validator adds a message to the messages array using [Array.prototype.push](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/push). The documentation for that method uses similar wording ("**adds** one or more elements to the end of an array"), and it's simpler and clearer than "extend", which makes me think of [jQuery.extend](https://api.jquery.com/jquery.extend/).

Secondly, "according error" sounds very unnatural and weird (it might not be grammatically incorrect, since according can be an adjective, but it's not the right word to use here). "Appropriate error" or "relevant error" or "correct error" is much better and clearer.

Comment (arogachev) [inline tests/js/assets/js/yii.validation.test.js]: @nkovacs. Sorry, I thought you are suggesting to use one generic message instead of two. Yes, I can modify message to be more precise.

Comment (nkovacs) [inline tests/js/assets/js/yii.validation.test.js]: Ah, I see. Thanks!

**Comment by cebe** [comment]
Thanks for your effort!

> Sinon can not be installed via Composer Asset Plugin. If I add line "npm-asset/sinon": "^1.17.6" to required-dev section of composer.json and run composer update I get the following error:

adding it to composer.json worked fine for me. Seems like you have not set up your SSH public key in the github settings.

> @arogachev,@cebe, it is already possible to execute the tests in Travis?

I guess since it is running in a browser right now, we can not easily run this on travis directly.
I am for getting this PR ready first and then tackle the problem of automating the test run on travis.

I'd also recommend to limit the scope of this PR to yii.validation.js and add more tests in separate PRs.

**Comment by cebe** [comment]
I have edited the issue on https://github.com/yiisoft/yii2/issues/12840 to include a list of separete steps that should be made in separate PRs.

**Comment by cebe** [comment]
@arogachev I have sent you a pull request: https://github.com/arogachev/yii2/pull/1 if you merge it that commit should be inluded here.

**Comment by arogachev** [comment]
Implemented execution of tests using Node.js and NPM. PHP test app is not needed anymore.

**Comment by nkovacs** [comment]
There are two reasons I used php to run the tests and not node.
I wanted to avoid requiring node for Yii development, and that setup allowed me to use php code to generate the client options required by the js code (though that does technically make them integration tests).

**Comment by dynasource** [comment]
how do you test JS with php? Dont you still need a JS executioner like Phantom or Selenium to do the magic? If thats the case, it really doesnt matter to use Nodejs as an alternative, right?

About that fixture preparation. This can still be done? You run your PHP file to create the required JS fixtures?

**Comment by nkovacs** [comment]
You're right, that is a problem for Yii, because you don't have a php server in your CI setup.
We already had a running apache server with php in our CI setup (for acceptance tests) and we used [gulp-mocha-phantomjs](https://www.npmjs.com/package/gulp-mocha-phantomjs) (because we also already had a gulp setup).
Also, when you're working on something, you can just open the test page in a browser. That also gives you easy access to the browser's development tools and a nicer output.

**Comment by arogachev** [comment]
I think JS unit tests should not be dependent on PHP in any way.

As for `clientValidateAttribute` - we can test that correct JS code part is generated on PHP side. Also additionally we can have some integration tests with simple app, but this is out of scope of this task in my opinion.

As for Node.js vs PhantomJS for running tests - there is no exact answer which one to use or which one is better. Both have pros and cons. I decided to use Node.js.

Running tests in the browser with existing Node.js / NPM implementation doesn't make much sense. In case of errors you can see the place where it's coming from in stack trace. Also you can perform some debugging using same `console.log()` and `process.exit()` if needed. Full output can be found in detailed Travis CI build info (you push your code to development branch anyway). You can scroll and investigate it in details.

I don't think we should support both ways of executing tests (browser and command line), in case of PhantomJS it probably would be easier, but with Node.js you can't load testable code in usual manner.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: @cebe It's related. I updated PR's first post. Please see "Integration with Travis CI" section.

**Comment by nkovacs** [comment]
Using a browser to run the tests gives you access to much better debugging tools than a simple `console.log` and the stack trace.
It also allows you to test and debug the code in different browsers and you can even set up CI with different browsers, e.g. with browserstack.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: Switch from `precise` to `trusty` is required mainly because of `jsdom` dependency. It requires Node.js > 4. Precise has very old Node.js and NPM versions and they can not be updated because of default versions of C++ compilers. While it's possible somehow to install Node 6 on Precise, see this [document](https://github.com/nodesource/distributions/blob/master/OLDER_DISTROS.md), I think this is not elegant solution and I'm not 100% sure it will work correctly with Travis and container-based precise dist.

Trusty has v6.3.0 which is enough and way better and newer.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: Default version of HHVM on Precise is 3.6.6, on Trusty - 3.15.2, see https://github.com/yiisoft/yii2/pull/12971#issuecomment-260134653. That causes failing of 2 tests(#12981, #12982). By specifying exact version we keep the same version as on Precise until these 2 bugs will be fixed. In this case tests are passing without failures.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: Additionally check Node.js and NPM versions.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: Install npm dev dependencies used for testing.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: Run the actual JS unit tests.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: Using `root` user fails:

```
yii\db\Exception: SQLSTATE[HY000] [1045] Access denied for user 'root'@'localhost' (using password: NO)
```
Using `travis` user fails too:

```
yii\db\Exception: SQLSTATE[HY000] [1045] Access denied for user 'travis'@'localhost' (using password: NO)
```

See also #12971.

So I decided to create separate dedicated user for test.

Also note that trusty by default uses MySQL 5.6, while precise - 5.5.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: This is a temporary workaround, see #12987.

**Comment by arogachev** [comment]
Please review and give some advices on solving #12987.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: This is an encrypted token from my account. I guess it's better to replace it with token of someone from core developers.

File: .travis.yml
Comment (arogachev) [inline .travis.yml]: This needs to be added:

```yaml
before_install:
  - composer config -g github-oauth.github.com "$GITHUB_TOKEN"
```

See https://github.com/yiisoft/yii2/issues/12987#issuecomment-260254948.

**Comment by arogachev** [comment]
#12987 is currently blocking this from successfully passing tests. Without it overall tests pass without failing.

**Comment by arogachev** [comment]
@nkovacs While I agree with you that testing with browser / different browsers has some advantages, here is my opinion.

> Using a browser to run the tests gives you access to much better debugging tools than a simple console.log and the stack trace.

How often this is actually needed? For existing functionality I think `console.log` and stack trace are enough. For something complex or new functionality we can generate test application. But we can add support for running tests via both command line and browser. Need others opinions.

> It also allows you to test and debug the code in different browsers and you can even set up CI with different browsers, e.g. with browserstack.

I checked Browserstack, It seems to be a good tool, but:

- Yii2 is not a JS framework.
- It has relatively small amount of JS code.
- JS code uses ES5 and some external libraries like JQuery, pjax, Punycode.js which are already tested.
- In unit tests stubs are used for testing some code, for example related with File API, jQuery, AJAX. To simulate real browser behavior we need integration tests which are good to have but unit tests have higher priority in my opinion.
- Browserstack is a commercial service. Even it's free for open source, as I understood, you need to contact them with details about your project before using. We don't know their conditions, maybe he will refuse, because of specifics of Yii2 code - and it's coupled with PHP framework and can't be used outside of it.
- We need to start with something and move from dead point. We can have this setup and consider using this later. I tried to not pollute whole file with Node.js integration.

**Comment by nkovacs** [comment]
I don't think it'll be that difficult to set up both browser and nodejs testing, since mocha supports both. I definitely think you should have a proper in-browser CI setup (not necessarily browserstack). It doesn't matter that you have very little JS code, one small thing like the quoted boolean above would make the entire code break in old IE for example (and you don't need integration tests to catch that).

But yeah, I agree, it's lower priority.

File: package.json
Comment (cebe) [inline package.json]: is this "version" key needed? I'd remove it to avoid the need of maintaining it.

**Comment by cebe** [comment]
moved to #13111, see PR description of #13111 for more details.

**Pull request closed without merging**
